Source/Core/AI/ThinkNodes/ThinkNode_Labor.cs

```diff
@@ -1,59 +1,308 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using PrisonLabor.Core.LaborArea;
+using PrisonLabor.Core.LaborWorkSettings;
+using PrisonLabor.Core.Meta;
 using PrisonLabor.Core.Needs;
 using PrisonLabor.Core.Other;
-using PrisonLabor.Core.Trackers;
 using RimWorld;
+using UnityEngine.Assertions.Must;
 using Verse;
 using Verse.AI;
 
 namespace PrisonLabor.Core.AI.ThinkNodes
 {
-    internal class ThinkNode_Labor : ThinkNode_Conditional
+    public class ThinkNode_Labor : ThinkNode
     {
-        protected override bool Satisfied(Pawn pawn)
+        public bool emergency;
+
+        public object Tutorials { get; private set; }
+
+        private readonly int jobsPerTick = 7;
+
+        private static TickManager manager = null;
+
+        private static Dictionary<Pawn, int> lastIndex = new Dictionary<Pawn, int>();
+
+        public override ThinkResult TryIssueJobPackage(Pawn pawn, JobIssueParams jobParams)
         {
-            if (pawn.IsPrisoner)
+            var mneed = pawn.needs.TryGetNeed<Need_Motivation>();
+
+            // Check for medical needs
+            if (HasMedicalNeed(pawn))
+                return SetNotWorking(mneed);
+
+            // Check the timetables in the worktab 
+            if (!HasWorkTimeAssignment(pawn))
+                return SetNotWorking(mneed);
+
+            // Check if the motivation is enabled
+            if (PrisonLaborPrefs.EnableMotivationMechanics && (mneed == null || mneed.IsLazy))
+                return SetNotWorking(mneed);
+
+            // Setting the tickmanager
+            if (manager == null) { manager = Find.TickManager; }
+
+            // TODO Remove this..
+            // Work prisoners will do]
+            if (pawn.timetable == null)
             {
-                //show tutorials
-                Tutorials.Introduction();
-                Tutorials.Management();
+                WorkSettings.InitWorkSettings(pawn);
+
+                // TODO: improve locks
+                // Restrict pawn to labor area
+                pawn.playerSettings.AreaRestriction = pawn.Map.areaManager.Get<Area_Labor>();
+            }
 
-                IntVec3 c;
+            // Parssing the entier job list "those selected in work tab"
+            var workList = pawn.workSettings.WorkGiversInOrderNormal;
+            var targetInfo = TargetInfo.Invalid;
 
-                var need = pawn.needs.TryGetNeed<Need_Motivation>();
-                if (need == null)
+            Job job = null;
+            WorkGiver_Scanner scanner = null;
+
+            // Checking if the dict contain the last time this pawn ticked
+            if (!lastIndex.ContainsKey(pawn)) { lastIndex.Add(pawn, 0); }
+
+            if (lastIndex[pawn] >= workList.Count) { lastIndex[pawn] = 0; }
+
+            // testing each job
+            var limit = lastIndex[pawn] + jobsPerTick;
+            for (var j = lastIndex[pawn]; j < workList.Count && j < limit; j++)
```

Comment:
This will mess with work priority when pawn has more jobs than 7. In one tick he will check high priority jobs in another low piority. In thight area it doesnt really matter but in large base when prisoner do for example cleaning duty it may force them to go to another end of base.

Reply:
OK, did you try this build or not cuz if you have a better faster solution i'm all ears. (the old one is spiking up to 30ms which is unacceptable)